### PR TITLE
drivers: ethernet: Deprecate eth_mcux, and some improvements to nxp_enet

### PR DESCRIPTION
**Pull request by decsny**
See commit messages for descriptions. 

Note that this PR contains a commit cherry-picked from #69207. This is because the ksz8081 phy driver improvements are necessary to unmark the new driver as experimental, because this PHY is on more or less all of the NXP EVKs - and some of the changes are crucial functional fixes.

### CODE REVIEW

File: drivers/ethernet/nxp_enet/Kconfig
Comment (danieldegrasse) [inline drivers/ethernet/nxp_enet/Kconfig:29]: We should add something to the release notes highlighting this change

File: drivers/ethernet/nxp_enet/eth_mcux.c
Comment (danieldegrasse) [inline drivers/ethernet/nxp_enet/eth_mcux.c:1]: I get the reasoning behind renaming the driver and causing a build error for users, but I think the point of `CONFIG_DEPRECATED` is to help users transition. IMO we should keep Kconfigs in the old driver as is

File: boards/nxp/mimxrt1160_evk/dts/nxp,enet-experimental.overlay
Comment (danieldegrasse) [inline boards/nxp/mimxrt1160_evk/dts/nxp,enet-experimental.overlay:1]: We should remove the testcase yaml cases as well now, since they are no longer needed

File: drivers/ethernet/nxp_enet/Kconfig
Comment (dleach02) [reply drivers/ethernet/nxp_enet/Kconfig:29]: agreed. Need to add to the release notes and probably the migration guide (maybe)

File: drivers/ethernet/eth_nxp_enet.c
Comment (decsny) [reply drivers/ethernet/eth_nxp_enet.c:782]: I agree, I deleted it on accident, but, it needs to have a new DT property introduced then because, as discussed on some other open tickets to ethernet drivers, this behavior is not part of the common zephyr/DT API, it either needs to be documented that the default behavior if a node has no Mac address is to use a pretermined one, or there needs to be another explicit DT property to clarify it. Right now a lot of drivers leave this to completely undefined behavior or even explicitly statically assert this case and cause a build error if the mac address is not specified in DT.

So the previous behavior is not standard and cannot be kept as it is... but I will definitely address adding the feature somewhat properly in this PR. Although I will not take any approach that changes the common ethernet interface in DT or documentation regarding this in this PR as that would require ensuring it to be true for all other ethernet drivers which is a scope creep.

File: drivers/ethernet/nxp_enet/Kconfig
Comment (decsny) [reply drivers/ethernet/nxp_enet/Kconfig:29]: Yes I was thinking about this I just haven't got to it yet (PR has a few loose ends I need to address in the morning)

File: drivers/ethernet/nxp_enet/eth_mcux.c
Comment (decsny) [reply drivers/ethernet/nxp_enet/eth_mcux.c:1]: Okay, I knew this would be controversial and I don't really care that much so I will just not rename them. 

Although I have to state an observation:

Small build errors from successive versions of zephyr are purposefully introduced all the time. The only reason we seem to want to make this depreciation a seamless transition is because the change is somewhat bigger, and therefore scarier. There appears to be an counterintuitive *inverse relationship* between the magnitude of a change being made and the amount of disruption produced by that change that can be tolerated. 

An opinion about this. Part of this is because we *decided* to do this this way in this case. It's actually very common for us to rewrite drivers and completely change DT binding schema and rename kconfigs. All the time in fact... to name a few: display drivers, LPADC, the PIT counter... but with this we decided it is different, even though it is not really. The real truth here (on the overall introduction of a second driver) is that the only obvious difference from a user perspective is configuration options, just like any other change we regularly make.

So my only point here is that I think that should be consistent in how we present changes to users. Whether we do it with this deprecation strategy (for which this has been a good experiment), or the way we normally do it, to me it doesn't matter, clearly user feedback would be useful, but right now it seems the deciding factor on a course of action for the process of change is the magnitude of fear of the change, which, in my opinion, does not facilitate a valid strategy.

File: boards/nxp/mimxrt1160_evk/dts/nxp,enet-experimental.overlay
Comment (decsny) [reply boards/nxp/mimxrt1160_evk/dts/nxp,enet-experimental.overlay:1]: Indeed

File: drivers/ethernet/eth_nxp_enet.c
Comment (decsny) [reply drivers/ethernet/eth_nxp_enet.c:782]: To clarify, what I am going to just do for now in this PR is introduce a boolean option in the nxp,enet binding to accomplish this, in a followup PR I will investigate a standard way to do it since it is a common feature and this keeps repearedly coming up as a point of debate on other ethernet driver PRs as well

Comment (decsny) [reply drivers/ethernet/eth_nxp_enet.c:782]: Also, the fact that the old code is setting the LAA bit seems incorrect to me, since what we are literally doing is the description of a UAA address, ie, using the vendor OUI and then setting the rest based on some unique silicon id. The only controversy about is we are deciding this in OS software rather than simply reading the entire address from some ROM. I still think it's not right to set the LAA bit here, but I'm not sure.

File: drivers/ethernet/nxp_enet/eth_mcux.c
Comment (dleach02) [reply drivers/ethernet/nxp_enet/eth_mcux.c:1]: Deprecation is a guiding principle of Zephyr project. 

Step 1 (release N): Add an experimental feature to replace an existing feature
Step 2 (release N+1): Transition experimental feature to non-experimental and mark legacy as deprecated
Step 3 (release N+3): Retire legacy feature. Note that the deprecation rules state that something stays in deprecation state for two release cycles before retiring.

Each of these steps shall ensure that legacy feature continues  to work to allow the consumer to react on their time and  not ours. 

https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html

Comment (decsny) [reply drivers/ethernet/nxp_enet/eth_mcux.c:1]: That documentation is for APIs. The closest thing to an API that we are talking about is configuration options (DT and kconfig). Currently everything selecting the CONFIG_DEPRECATED symbol is either an actual API or some feature that is literally no longer supported. In this case, the feature is the same, the driver is just changing. We could easily have just presented it as we do with any other driver change. We force users to make configuration changes all of the time, such as the examples I listed above and plenty more. If it's the implementation that we're concerned about, then what do we have to say then about the massive HAL updates that we do every few months that change millions of lines of thousands of files, that no single reviewer can even nearly begin to comprehend in full? They aren't API changes to the user, so we don't do a deprecation process of every HAL version, because it's all implementation. How is this different? 

If we want to do it this way from now on for any implementation change or configuration option update, all I'm saying is we should be consistent. Don't understand exactly what is special about this case that it is being treated like an API.

**Comment by pdgendt** [comment]
Will the new nxp_enet driver work with Distributed Switch Architecture `DSA`?

From `drivers/ethernet/Kconfig.dsa`:
```
menuconfig NET_DSA
	bool "Distributed Switch Architecture support"
	depends on ETH_MCUX || ETH_SAM_GMAC || ETH_STM32_HAL
	help
	  Enable Distributed Switch Architecture support. For now it
	  only supports Kinetics and STM32 ENET drivers.
```

**Comment by decsny** [comment]
> Will the new nxp_enet driver work with Distributed Switch Architecture `DSA`?
> 
> From `drivers/ethernet/Kconfig.dsa`:
> 
> ```
> menuconfig NET_DSA
> 	bool "Distributed Switch Architecture support"
> 	depends on ETH_MCUX || ETH_SAM_GMAC || ETH_STM32_HAL
> 	help
> 	  Enable Distributed Switch Architecture support. For now it
> 	  only supports Kinetics and STM32 ENET drivers.
> ```

I should hopefully get the Segger IPk66f board soon to test it.... but it should have the same dsa related code as eth_mcux did

**Comment by decsny** [comment]
@pdgendt I got one of the segger ipk66f boards as I mentioned, and created a little switched network consisting of:
1. The ipk66f board using the new nxp_enet driver and acting as a switch using DSA
2. A linux PC on one of the dsa slave ports
3. frdm k64f running with the new nxp_enet driver on one of the slave ports
4. frdm mcx947 board running with the nxp_enet_qos driver that just merged

did some developer testing using this network to verify that the driver seems to be working for DSA, so added the commit to show support for it

File: drivers/ethernet/eth_nxp_enet.c
Comment (jukkar) [inline drivers/ethernet/eth_nxp_enet.c:107]: Perhaps `static` ?

File: drivers/ethernet/eth_nxp_enet.c
Comment (jukkar) [inline drivers/ethernet/eth_nxp_enet.c:436]: missing `{}` although this looks just fine too. I think there was some Misra rule for this, and a compliance checker script might complain about the missing braces.

Unrelated to this PR, I noticed while checking what `eth_nxp_enet_rx()` is doing, that it might make sense in that function to return the real error from `net_recv_data()` instead of always `-EIO`

**Review by jukkar — COMMENTED** [review]
Couple of minor nits, and the CI needs to be fixed, but overall LGTM

File: drivers/ethernet/Kconfig.dsa
Comment (pdgendt) [inline drivers/ethernet/Kconfig.dsa:9]: Thanks for this addition, however this doesn't really scale, and doesn't allow for OOT drivers.

Maybe we need to add a hidden kconfig, something like `ETH_SUPPORT_DSA`, which is selected by those drivers?

```
config ETH_DSA_SUPPORT
	bool
	help
	  Hidden symbol to indicate DSA is supported by an Ethernet driver.

menuconfig NET_DSA
	bool "Distributed Switch Architecture support"
	depends on ETH_DSA_SUPPORT
```

Comment (decsny) [reply drivers/ethernet/Kconfig.dsa]: okay, I added a commit converting this

File: drivers/ethernet/eth_nxp_enet.c
Comment (decsny) [reply drivers/ethernet/eth_nxp_enet.c]: WAIT_FOR didn't make sense to me to use because, we aren't waiting on something, ie there is no notion of time here, it's just repeat this routine until done, so I elected to convert it to a do-while with a return variable

File: doc/releases/migration-guide-3.7.rst
Comment (decsny) [reply doc/releases/migration-guide-3.7.rst]: I took this to mean that you think the word "ethernet" in this line should be capitalized, which I updated

File: drivers/ethernet/eth_nxp_enet.c
Comment (pdgendt) [inline drivers/ethernet/eth_nxp_enet.c:118]: Nit: align

**Review by pdgendt — DISMISSED** [review]

**Review by pdgendt — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Comment by kartben** [comment]
Re-assigning to @decsny - although he's the author of this PR, he's the most appopriate assignee in this particular instance

**Pull request merged**
